README.md
# Serverless Artifact Cacher

[![Build Status](https://travis-ci.org/troyready/serverless-artifact-cacher.svg?branch=master)](https://travis-ci.org/troyready/serverless-artifact-cacher)

This project deploys a [Serverless](https://serverless.com/cli/) artifact caching service. It currently supports transparent caching of:

* [npm](https://www.npmjs.com/) packages for Node.js
* [pip](https://pypi.org/) packages for Python

## Setup

### Pre-reqs

* Clone the project
* Configure the allowed IPs to access the service:
    * Search for `aws:SourceIp` in serverless.js & edit/uncomment the IP restriction, turning:

```
          // "Condition": {
          //   "IpAddress": {
          //     "aws:SourceIp": ["X.X.X.X"]
          //   }
          // }
```
into something like:
```
          "Condition": {
            "IpAddress": {
              "aws:SourceIp": ["1.2.3.4"]
            }
          }
```

Serverless [variable lookup](https://serverless.com/framework/docs/providers/aws/guide/variables/) can be used here to dynamically retrieve the values, e.g. `"aws:SourceIp": ["${ssm:/path/to/stringlistparam~split}"]`
 
### Deploying

* Run `npm install`
    * If any errors arise try deleting `package-lock.json` and trying again
* Run sls deploy for your stage & region; e.g. for the "common" stage in oregon: `npx sls deploy -s common -r us-west-2`

Upon completion, you can determine your cache URLs from the displayed endpoints:

* NPM:
    * `https://SOMEID.execute-api.REGION.amazonaws.com/STAGE/npm`
* PyPi:
    * `https://SOMEID.execute-api.REGION.amazonaws.com/STAGE/pypi`

These URLs can be programatically retrieved via the CloudFormation stack (stack name `artifact-cacher-STAGE`) output `ServiceEndpoint` -- append `/npm` or `pypi` to it for their respective services.

## Use

In your execution environment (e.g. CodeBuild w/ VPC configuration), perform the following:

### NPM

Set the `NPM_CONFIG_REGISTRY` environment variable to your NPM endpoint

### PyPi (Python)

#### Pipenv (Recommended)

For [Pipenv](https://github.com/pypa/pipenv), set the `PIPENV_PYPI_MIRROR` environment variable to your PyPi endpoint.

#### Pip

For regular pip installs, run:
```
mkdir -p ~/.pip/pip.conf
echo "[global]
index-url = YOURPYPIENDPOINT" >> ~/.pip/pip.conf
```
(substituting the PyPi endpoint for YOURPYPIENDPOINT)

## Operation/Maintenance

Each repository should be self-sufficient:
* Upon request for a new (not-before-cached) package, the service will cache the list of available versions for that package.
* Upon request to download a package version, the service will first cache the download and then offer the download from the cache

Subsequent installs of a package will be served entirely from the local cache.

### Automatic Updates

Requests for a previously retrieved package are served solely from the cache, so newly uploaded package versions won't be displayed. The `AutoUpdate` functions will run daily to incorporate new package versions in into the cached version list.

The schedule can be adjusted in `serverless.js` (search for `rate(24 hours)`), and the update process can be triggered at any time by invoking the AutoUpdate functions.

### Cached PIP Versions

The DynamoDB table for each repository stores an item for each cached package. New upstream entries will be automatically added daily (see "Automatic Updates")

#### Cached NPM Versions

The NPM data is stored gzip format. It can be downloaded from s3bucket://${packageName}/index.gz.
- Download the file and from the GUI click on "Decompress" in the Convert menu.
- From the command line type `gzip -kd index.gz`

New upstream entries will be automatically added daily (see "Automatic Updates")


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2020-04-30

### Changed

- NPM package data is now stored on S3, eliminating errors from entries too-large for DDB (like `typescript`)

## [1.0.0] - 2019-12-30

### Added

- Initial release

[Unreleased]: https://github.com/troyready/serverless-artifact-cacher/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/troyready/serverless-artifact-cacher/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/troyready/serverless-artifact-cacher/releases/tag/v1.0.0


serverless.js
async function getConfig() {
  const generalConfig = {
    "service": "artifact-cacher",
    "plugins": [
      "serverless-webpack",
      "serverless-iam-roles-per-function"
    ],
    "provider": {
      "name": "aws",
      "runtime": "nodejs10.x",
      "resourcePolicy": [
        {
          "Effect": "Allow",
          "Principal": "*",
          "Action": "execute-api:Invoke",
          "Resource": [
            "execute-api:/*/*/*"
          ],
          // "Condition": {
          //   "IpAddress": {
          //     "aws:SourceIp": ["X.X.X.X"]
          //   }
          // }
        }
      ]
    },
    // Packaging each function individually cuts down their size significantly,
    // but can cause memory issues
    // https://github.com/serverless-heaven/serverless-webpack/issues/299
    // To safely enable this, it's advised to:
    // 1) Run 'export NODE_OPTIONS=--max_old_space_size=8192' or (windows) '$Env:NODE_OPTIONS = "--max_old_space_size=8192"', and
    // 2) Apply the patch from https://github.com/serverless-heaven/serverless-webpack/pull/517 to serverless-webpack
    //    (just update the lib/Configuration.js, lib/compile.js, & lib/validate.js files)
    // "package": {
    //   "individually": true
    // },
    "custom": {
      "webpack": {
        "excludeFiles": [
          "src/**/*.test.ts",
          "src/**/__mocks__/*.ts"
        ]
        // We could exclude aws-sdk here to trim down the deployment package via
        // webpack-node-externals (see webpack.config.js), but instead we'll
        // include it so the sdk version used is predictable.
        // "includeModules": {
        //   "forceExclude": [
        //     "aws-sdk"
        //   ]
        // }
      }
    }
  };

  const npmResources = {};
  const npmStorageBucketResourceName = "NpmStorageBucket";
  npmResources[npmStorageBucketResourceName] = {
    "Type": "AWS::S3::Bucket",
    "DeletionPolicy": "Retain"
  }

  const npmFunctions = {
    "npmHyphen": {
      "handler": "src/npm/hyphen/hyphen.handler",
      "events": [
        {
          "http": {
            "path": "npm/-/{proxy+}",
            "method": "get",
            "integration": "lambda-proxy"
          }
        }
      ]
    },

    "npmPackage": {
      "handler": "src/npm/package/package.handler",
      "environment": {
        "BUCKET_NAME": {
          "Ref": npmStorageBucketResourceName
        },
        "NPM_CACHE_DOWNLOAD_URI": {
          "Fn::Join": [
            "",
            ["https://",
             {"Ref": "ApiGatewayRestApi"},
             ".execute-api.",
             {"Ref": "AWS::Region"},
             ".",
             {"Ref": "AWS::URLSuffix"},
             "/${opt:stage}/npm-dlredirect"]
          ]
        }        
      },
      "iamRoleStatements": [
        {
          "Action": [
            "s3:GetObject",
            "s3:PutObject"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::GetAtt": [
                    npmStorageBucketResourceName,
                    "Arn"
                  ]
                },
                "/*"
              ]
            ]
          }
        }
      ],
      "events": [
        {
          "http": {
            "path": "npm/{proxy+}",
            "method": "get",
            "integration": "lambda-proxy"
          }
        }
      ]
    },

    "npmAutoUpdate": {
      "handler": "src/npm/autoupdate/autoupdate.handler",
      "timeout": 900,
      "environment": {
        "BUCKET_NAME": {
          "Ref": npmStorageBucketResourceName
        },
        "NPM_CACHE_DOWNLOAD_URI": {
          "Fn::Join": [
            "",
            ["https://",
             {"Ref": "ApiGatewayRestApi"},
             ".execute-api.",
             {"Ref": "AWS::Region"},
             ".",
             {"Ref": "AWS::URLSuffix"},
             "/${opt:stage}/npm-dlredirect"]
          ]
        }
      },
      "iamRoleStatements": [
        {
          "Action": [
            "s3:ListBucket"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::GetAtt": [
              npmStorageBucketResourceName,
              "Arn"
            ]
          }
        },
        {
          "Action": [
            "s3:GetObject",
            "s3:PutObject"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::GetAtt": [
                    npmStorageBucketResourceName,
                    "Arn"
                  ]
                },
                "/*"
              ]
            ]
          }
        }
      ],
      "events": [
        {
          "schedule": "rate(24 hours)"
        }
      ]
    },

    "npmDlRedirect": {
      "handler": "src/npm/dlredirect/dlredirect.handler",
      "timeout": 30,
      "environment": {
        "BUCKET_NAME": {
          "Ref": npmStorageBucketResourceName
        }
      },
      "iamRoleStatements": [
        {
          "Action": [
            "s3:ListBucket"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::GetAtt": [
              npmStorageBucketResourceName,
              "Arn"
            ]
          }
        },
        {
          "Action": [
            "s3:GetObject",
            "s3:PutObject"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::GetAtt": [
                    npmStorageBucketResourceName,
                    "Arn"
                  ]
                },
                "/*"
              ]
            ]
          }
        }
      ],
      "events": [
        {
          "http": {
            "path": "npm-dlredirect/{proxy+}",
            "method": "get",
            "integration": "lambda-proxy"
          }
        }
      ]
    }
  };

  const pypiResources = {};
  const pypiStorageBucketResourceName = "PyPiStorageBucket";
  pypiResources[pypiStorageBucketResourceName] = {
    "Type": "AWS::S3::Bucket",
    "DeletionPolicy": "Retain"
  }
  const pypiTableResourceName = "PyPiTable";
  pypiResources[pypiTableResourceName] = {
    "Type": "AWS::DynamoDB::Table",
    "Properties": {
      "AttributeDefinitions": [
        {
          "AttributeName": "PackageName",
          "AttributeType": "S"
        }
      ],
      "KeySchema": [
        {
          "AttributeName": "PackageName",
          "KeyType": "HASH"
        }
      ],
      "BillingMode": "PAY_PER_REQUEST"
    }
  };

  const pypiFunctions = {
    "pypiList": {
      "handler": "src/pypi/list/list.handler",
      "environment": {
        "DDB_TABLE": {
          "Ref": pypiTableResourceName
        }
      },
      "iamRoleStatements": [
        {
          "Action": [
            "dynamodb:Scan"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::GetAtt": [
              pypiTableResourceName,
              "Arn"
            ]
          }
        }
      ],
      "events": [
        {
          "http": {
            "path": "pypi/",
            "method": "get",
            "integration": "lambda-proxy"
          }
        }
      ]
    },

    "pypiPackage": {
      "handler": "src/pypi/package/package.handler",
      "environment": {
        "DDB_TABLE": {
          "Ref": pypiTableResourceName
        }
      },
      "iamRoleStatements": [
        {
          "Action": [
            "dynamodb:GetItem",
            "dynamodb:PutItem"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::GetAtt": [
              pypiTableResourceName,
              "Arn"
            ]
          }
        }
      ],
      "events": [
        {
          "http": {
            "path": "pypi/{proxy+}",
            "method": "get",
            "integration": "lambda-proxy"
          }
        }
      ]
    },

    "pypiAutoUpdate": {
      "handler": "src/pypi/autoupdate/autoupdate.handler",
      "timeout": 900,
      "environment": {
        "DDB_TABLE": {
          "Ref": pypiTableResourceName
        }
      },
      "iamRoleStatements": [
        {
          "Action": [
            "dynamodb:PutItem",
            "dynamodb:Scan"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::GetAtt": [
              pypiTableResourceName,
              "Arn"
            ]
          }
        }
      ],
      "events": [
        {
          "schedule": "rate(24 hours)"
        }
      ]
    },

    "pypiDlRedirect": {
      "handler": "src/pypi/dlredirect/dlredirect.handler",
      "timeout": 30,
      "environment": {
        "BUCKET_NAME": {
          "Ref": pypiStorageBucketResourceName
        }
      },
      "iamRoleStatements": [
        {
          "Action": [
            "s3:ListBucket"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::GetAtt": [
              pypiStorageBucketResourceName,
              "Arn"
            ]
          }
        },
        {
          "Action": [
            "s3:GetObject",
            "s3:PutObject"
          ],
          "Effect": "Allow",
          "Resource": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::GetAtt": [
                    pypiStorageBucketResourceName,
                    "Arn"
                  ]
                },
                "/*"
              ]
            ]
          }
        }
      ],
      "events": [
        {
          "http": {
            "path": "pypi-dlredirect/{proxy+}",
            "method": "get",
            "integration": "lambda-proxy"
          }
        }
      ]
    }
  };

  return {
    ...generalConfig,
    "functions": {
      ...npmFunctions,
      ...pypiFunctions
    },
    "resources": {
      "Resources": {
        ...npmResources,
        ...pypiResources
      }
    }
  };
}

module.exports = getConfig();
